lib/drivers/bayerContourNext.js

```diff
@@ -470,48 +514,63 @@ module.exports = function (config) {
     fetchData: function (progress, data, cb) {
       debug('in fetchData', data);
 
-      function getOneRecordWithProgress(recnum, cb) {
-        var rec = data.rawrecords.shift();
-        progress(100.0 * recnum / data.nrecs);
-        setTimeout(function() {
-          getOneRecord(rec, data, cb);
-        }, 20);
-      }
-
-      async.timesSeries(data.nrecs, getOneRecordWithProgress, function(err, result) {
-        if (err) {
-          debug('fetchData failed');
-          debug(err);
-          debug(result);
-        } else {
-          debug('fetchData', result);
+      var recordType = null;
+      var dataRecords = [];
+      var error = false;
+
+      async.whilst(
+        // Get records from the meter until we get the Message Terminator Record (L)
+        // The spec says that unless we get this, any preceding data should not be used.
+        function () { return (recordType !== 'L' && !error); },
+        function (callback) {
+          getOneRecord(data, function (err, result) {
+            recordType = result.recordType;
```

Comment:
If `getOneRecord` returns an error, this code will still try to process `result` first before handling the error. Also, remember to do a `return callback(err,null)`, with just a `callback(err,null)` it will still try to process and upload the invalid data.